lib/runner.js

```diff
@@ -162,6 +168,53 @@ Runner.immediately = global.setImmediate || process.nextTick;
  */
 inherits(Runner, EventEmitter);
 
+/**
+ * Replacement for `target.on(eventName, listener)` that does bookkeeping to remove them when this runner instance is disposed.
+ * @param target {EventEmitter}
+ * @param eventName {string}
+ * @param fn {function}
+ */
+Runner.prototype._addEventListener = function(target, eventName, listener) {
+  target.on(eventName, listener);
+  this._eventListeners.push([target, eventName, listener]);
+};
+
+/**
+ * Replacement for `target.removeListener(eventName, listener)` that also updates the bookkeeping.
+ * @param target {EventEmitter}
+ * @param eventName {string}
+ * @param fn {function}
+ */
+Runner.prototype._removeEventListener = function(target, eventName, listener) {
+  var eventListenerIndex = this._eventListeners.findIndex(function(
```

Comment:
This looks to only be used for `process`, but it iterates over everything in `this._eventListeners`.  That doesn't look necessary.  Maybe instead we keep an array of _process_ event listeners (containing only `eventName` and `listener`, or an object with `eventName` key and an array of listener functions as value)--say, `this._processEventListeners`--and instead of naming this function `_removeEventListener`, call it `_removeProcessEventListener`.  In `dispose()`, iterate over this data structure in addition to `this._eventListeners`... assuming removing process listeners in `dispose()` is intended.

Reply:
It is also used for `'error'` listeners on tests and hooks. (L440 and L592)

Reply:
ah, didn't see that.  I am concerned that this is a CPU-bound bottleneck in very large test files / suites.  if it turns out to be, we can refactor it later to be more difficult to understand but more performant. :smile: